lib/logstash/inputs/http.rb

```diff
@@ -0,0 +1,100 @@
+# encoding: utf-8
+require "logstash/inputs/base"
+require "logstash/namespace"
+require "stud/interval"
+require "socket" # for Socket.gethostname
+require "puma/server"
+require "puma/minissl"
+
+class Puma::Server
+  # ensure this method doesn't mess up our vanilla request
+  def normalize_env(env, client); end
+end
+
+# This output opens a web server and listens to HTTP requests,
+# converting them to LogStash::Event instances.
+#
+class LogStash::Inputs::Http < LogStash::Inputs::Base
+  config_name "http"
+
+  # If undefined, Logstash will complain, even if codec is unused.
+  default :codec, "plain"
+
+  # Which host or ip to bind to
+  config :host, :validate => :string, :default => "0.0.0.0"
+
+  # Which TCP port to bind to
+  config :port, :validate => :number, :default => 8080
+
+  # Maximum number of threads to use
+  config :max_threads, :validate => :number, :default => 4
+
+  # SSL Configurations
+  #
+  # Enable SSL
+  config :ssl, :validate => :boolean, :default => false
+
+  # The JKS keystore to validate the client's certificates
+  config :keystore, :validate => :path
+
+  #TODO: config :cacert, :validate => :path
+
+  # Set the truststore password
+  config :keystore_password, :validate => :password
+
+  # useless headers puma adds to the requests
+  # mostly due to rack compliance
+  REJECTED_HEADERS = ["puma.socket", "rack.hijack?", "rack.hijack", "rack.url_scheme", "rack.after_reply", "rack.version", "rack.errors", "rack.multithread", "rack.multiprocess", "rack.run_once", "SCRIPT_NAME", "QUERY_STRING", "SERVER_PROTOCOL", "SERVER_SOFTWARE", "GATEWAY_INTERFACE"]
+
+  RESPONSE_HEADERS = {'Content-Type' => 'text/html'}
+
+  public
+  def register
+    @server = ::Puma::Server.new(nil) # we'll set the rack handler later
+    if @ssl
+      ctx = Puma::MiniSSL::Context.new
+      ctx.keystore = @keystore
+      ctx.keystore_pass = @keystore_password.value
+      @server.add_ssl_listener @host, @port, ctx
+    else
+      @server.add_tcp_listener @host, @port
+    end
+    @server.min_threads = 0
+    @server.max_threads = @max_threads
+  end # def register
+
+  def run(queue)
+    @server.app = Proc.new do |req|
+      begin
+        REJECTED_HEADERS.each {|k| req.delete(k) }
+        req = lowercase_keys(req)
+        body = req.delete("rack.input")
+        @codec.decode(body.read) do |event|
+          event["headers"] = req
+          decorate(event)
+          queue << event
+        end
+        ['200', RESPONSE_HEADERS, ['ok']]
+      rescue => e
+        @logger.error("unable to process event #{req.inspect}. exception => #{e.inspect}")
+        ['500', RESPONSE_HEADERS, ['internal error']]
+      end
```

Comment:
Thinking out loud here.

Since our http input is in reality a small webservice, we should consider the response format.

If we return a response to the callback we should use the same format as the request.
If they post a json document we should return a json document, same situation with xml.

To make things less complicated.. I dont think we need to return a string, the http code are enough in this case and more flexible.